### PR TITLE
Support arbitrary types via IComparer in source generator

### LINKED ISSUE #94: Source generator: support arbitrary types via IComparer (object/generic path)
## Problem

The old `NetworkSort` API had a generic `Sort<T>(T[], IComparer<T>)` overload that worked with any type — custom structs, records, etc. The source generator currently only supports 14 known types (13 primitives + string) and reports `SN0002` for anything else.

This means users can't do:

\\\csharp
[SortingNetwork(27, typeof(MyRecord))]  // SN0002 error
partial class MySorter { }
\\\

## Proposed solution

Add an `object` (or generic) implementation path in the generator. When `typeof(T)` is not one of the known primitive/string types, emit a **comparer-only** sort method that uses `IComparer<T>` for all comparisons (no scalar `>` operator, no SIMD):

\\\csharp
// Generated for [SortingNetwork(27, typeof(MyRecord))]
public static void Sort(Span<MyRecord> span, IComparer<MyRecord>? comparer = null)
{
    comparer ??= Comparer<MyRecord>.Default;
    // Apply network with comparer (loop-based, same as ApplyNetworkWithComparer)
}
\\\

This would use the existing `ApplyNetworkWithComparer` pattern — a loop over the network's comparator pairs calling `comparer.Compare(a, b)`.

## What this enables

- Sorting networks for custom types (structs, records, classes)
- Feature parity with the old `NetworkSort.Sort<T>(T[], IComparer<T>)` API
- The `comparer` parameter could default to `Comparer<T>.Default` for types implementing `IComparable<T>`

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Fixes #94

## Summary

When `typeof(T)` is not a known primitive or string, the source generator now emits a **comparer-only** sort path instead of reporting the `SN0002` error. This enables sorting networks for custom structs, records, and classes.

## What changed

**Generator (`SortingNetworkGenerator.cs`):**
- Removed the `SN0002` rejection for unknown types
- Added `IsCustomType` property to `NetworkRequest` 
- Custom types get `Sort(Span<T>, IComparer<T>? comparer = null)` and `Sort(T[], IComparer<T>? comparer = null)` -- the default-null comparer falls back to `Comparer<T>.Default`
- No parameterless `Sort(Span<T>)`, no scalar unrolled methods, no SIMD for custom types
- Uses `FullyQualifiedFormat` (`global::Namespace.Type`) so custom types resolve without extra `using` directives
- Known primitive/string types are completely unaffected

**Tests:**
- `TestTypes.cs` -- defines a `Point` record struct implementing `IComparable<Point>`
- `GeneratedSorters.cs` -- registers `Point` at sizes 4, 27, 28
- `GeneratedSortTests.cs` -- 14 new tests: correctness, stress (100 seeds), reverse comparer, null comparer, array overload, wrong size, edge cases
- `GeneratorTests.cs` -- 3 new tests: verifies generated code shape, namespaced types, mixed primitive+custom on same class

All 313 tests pass (296 existing + 17 new).

## Example usage

```csharp
public record struct MyRecord(int X, int Y) : IComparable<MyRecord> { ... }

[SortingNetwork(27, typeof(MyRecord))]
partial class MySorter { }

// Generated API:
MySorter.Sort(mySpan);                    // uses Comparer<MyRecord>.Default
MySorter.Sort(mySpan, myCustomComparer);  // uses provided comparer
MySorter.Sort(myArray);                   // array overload
```

### CODE REVIEW

File: SortingNetworks.Tests/GeneratedSortTests.cs [hunk 1]
```diff
@@ -1233,4 +1233,146 @@ public void Sort_NullComparer_UsesDefault_String()
...
+            var rng = new Random(seed);
+            var input = Enumerable.Range(0, size).Select(_ => RandomPoint(rng)).ToArray();
+            var expected = (Point[])input.Clone();
+            Array.Sort(expected);
+
+            var actual = (Point[])input.Clone();
+            GeneratedSorters.Sort(actual.AsSpan());
+
+            Assert.Equal(expected, actual);
+        }
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks.Tests/GeneratedSortTests.cs]: Good catch - refactored to use the existing StressSort helper in 990bea1.

File: SortingNetworks.Generators/SortingNetworkGenerator.cs [hunk 2]
```diff
@@ -542,10 +539,12 @@ private static void Execute(SourceProductionContext context, ImmutableArray<Gene
...
+                } // end if (!isCustomType)
 
                 // Emit comparer Span overload
+                string comparerDefault = isCustomType ? " = null" : "";
                 sb.AppendLine($"        /// <summary>Sorts a span of {typeName} using an optimal sorting network with a custom comparer.</summary>");
-                sb.AppendLine($"        public static void Sort(System.Span<{typeName}> span, System.Collections.Generic.IComparer<{typeName}>? comparer)");
+                sb.AppendLine($"        public static void Sort(System.Span<{typeName}> span, System.Collections.Generic.IComparer<{typeName}>? comparer{comparerDefault})");
                 sb.AppendLine("        {");
                 sb.AppendLine($"            comparer ??= System.Collections.Generic.Comparer<{typeName}>.Default;");
                 sb.AppendLine("            int n = span.Length;");
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks.Generators/SortingNetworkGenerator.cs:662]: Comparer<T>.Default doesn't throw on access — only when .Compare() is called on a type that doesn't implement IComparable<T>. Since the compare only happens inside the matched-length branch, the ordering here is fine. This is also the same pattern used for all known types.

File: SortingNetworks.Generators/SortingNetworkGenerator.cs [hunk 3]
```diff
@@ -278,6 +271,10 @@ private static void Execute(SourceProductionContext context, ImmutableArray<Gene
                 var sizes = kvp.Value;
                 sizes.Sort((a, b) => a.Size.CompareTo(b.Size));
 
+                bool isCustomType = sizes[0].IsCustomType;
+
+                if (!isCustomType)
+                {
                 // Check which sizes have SIMD support
                 var simdSizes = new List<NetworkRequest>();
                 var avx2FallbackSizes = new List<NetworkRequest>();
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks.Generators/SortingNetworkGenerator.cs]: Fixed in 99b1306 — flipped to early-return with continue for custom types, so the known-type block stays at its original indentation. Also extracted the shared comparer emission into an EmitComparerOverloads helper.
